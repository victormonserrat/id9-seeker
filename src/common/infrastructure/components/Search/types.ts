import { InputHTMLAttributes } from 'react'

export type SearchProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  'onChange' | 'value'
> & {
  onChange?: (value: string) => void
  value?: string
}

export type UseSearchProps = {
  onChange: SearchProps['onChange']
  value: SearchProps['value']
}
