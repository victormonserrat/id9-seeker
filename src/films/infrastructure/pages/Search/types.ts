import { FilmCardProps } from '~/films/infrastructure/components/Card'

export type FilmSearchProps = Readonly<{
  films: (FilmCardProps & { href: string; id: string })[]
}>
