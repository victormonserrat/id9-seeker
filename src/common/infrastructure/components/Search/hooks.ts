import { ChangeEvent, FormEvent, useCallback, useEffect, useState } from 'react'

import { UseSearchProps } from './types'

export const useSearch = ({ onChange, value }: UseSearchProps) => {
  const [inputValue, setInputValue] = useState(value ?? '')

  const handleChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => setInputValue(event.target.value),
    [],
  )

  const handleSubmit = useCallback(
    (event: FormEvent<HTMLFormElement>) => {
      if (!onChange) return

      event.preventDefault()
      onChange(inputValue)
    },
    [inputValue, onChange],
  )

  useEffect(() => setInputValue(value ?? ''), [value])

  return { handleChange, handleSubmit, inputValue }
}
