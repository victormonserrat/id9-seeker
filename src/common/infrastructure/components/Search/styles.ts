import styled from 'styled-components'

import { leftEdgeStyles } from '~/common/infrastructure/styles/edges'
import media from '~/common/infrastructure/styles/media'

const BUTTON_IMAGE_SIZE = 50
const BUTTON_SIZE = 42
const BUTTON_HEIGHT_MEDIUM = 46
const BUTTON_WIDTH_MEDIUM = 140
const BUTTON_IMAGE_OFFSET_X = (BUTTON_SIZE - BUTTON_IMAGE_SIZE) / 2
const BUTTON_IMAGE_OFFSET_X_MEDIUM =
  (BUTTON_HEIGHT_MEDIUM - BUTTON_IMAGE_SIZE) / 2
const BUTTON_IMAGE_OFFSET_Y = BUTTON_IMAGE_OFFSET_X - BUTTON_IMAGE_SIZE
const BUTTON_IMAGE_OFFSET_Y_MEDIUM =
  BUTTON_IMAGE_OFFSET_X_MEDIUM - BUTTON_IMAGE_SIZE
const INPUT_HEIGHT = 64
const INPUT_HEIGHT_MEDIUM = 68
const BUTTON_OFFSET = (INPUT_HEIGHT - BUTTON_SIZE) / 2
const BUTTON_OFFSET_MEDIUM = (INPUT_HEIGHT_MEDIUM - BUTTON_HEIGHT_MEDIUM) / 2
const INPUT_PADDING = 18
const INPUT_PADDING_MEDIUM = 26
const INPUT_PADDING_RIGHT = INPUT_PADDING + BUTTON_SIZE + BUTTON_OFFSET
const INPUT_PADDING_RIGHT_MEDIUM =
  INPUT_PADDING_MEDIUM + BUTTON_WIDTH_MEDIUM + BUTTON_OFFSET_MEDIUM

export const Button = styled.button`
  background-color: ${({ theme }) =>
    theme.colors.search.button.background.default};
  background-image: url(https://static-mh.content.disney.io/starwars/assets/shared/search/search_icon-2c79f0615f1d.png);
  background-position: ${BUTTON_IMAGE_OFFSET_X}px ${BUTTON_IMAGE_OFFSET_Y}px;
  background-repeat: no-repeat;
  border-radius: 0.375rem;
  color: ${({ theme }) => theme.colors.search.button.text};
  height: ${BUTTON_SIZE / 16}rem;
  overflow: hidden;
  position: absolute;
  right: ${BUTTON_OFFSET / 16}rem;
  text-indent: 100%;
  top: ${BUTTON_OFFSET / 16}rem;
  width: ${BUTTON_SIZE / 16}rem;

  :enabled:hover {
    background-color: ${({ theme }) =>
      theme.colors.search.button.background.hover};
  }

  ${media.medium} {
    background-position: ${BUTTON_IMAGE_OFFSET_X_MEDIUM}px
      ${BUTTON_IMAGE_OFFSET_Y_MEDIUM}px;
    font-size: ${({ theme }) => theme.font.sizes.large};
    font-weight: ${({ theme }) => theme.font.weights.light};
    height: ${BUTTON_HEIGHT_MEDIUM / 16}rem;
    padding-left: ${BUTTON_HEIGHT_MEDIUM / 16}rem;
    padding-right: 1rem;
    right: ${BUTTON_OFFSET_MEDIUM / 16}rem;
    text-indent: unset;
    text-overflow: ellipsis;
    text-transform: uppercase;
    top: ${BUTTON_OFFSET_MEDIUM / 16}rem;
    width: ${BUTTON_WIDTH_MEDIUM / 16}rem;
  }
`

export const Component = styled.form`
  background-color: ${({ theme }) => theme.colors.search.background};
  border-radius: 0.375rem;
  position: relative;

  ${media.medium} {
    ${({ theme }) => leftEdgeStyles(theme.colors.search.background)};
  }
`

export const Input = styled.input`
  color: ${({ theme }) => theme.colors.search.input.text};
  font-size: ${({ theme }) => theme.font.sizes.extralarge};
  font-weight: ${({ theme }) => theme.font.weights.extralight};
  height: ${INPUT_HEIGHT / 16}rem;
  padding: ${INPUT_PADDING / 16}rem;
  padding-right: ${INPUT_PADDING_RIGHT / 16}rem;
  width: 100%;

  ::placeholder {
    color: ${({ theme }) => theme.colors.search.input.placeholder};
  }

  ${media.medium} {
    height: ${INPUT_HEIGHT_MEDIUM / 16}rem;
    padding: ${INPUT_PADDING_MEDIUM / 16}rem;
    padding-right: ${INPUT_PADDING_RIGHT_MEDIUM / 16}rem;
  }
`

export const InputContainer = styled.div`
  ::after {
    background-color: ${({ theme }) => theme.colors.search.underline};
    bottom: 0.5rem;
    content: '';
    height: 0.0625rem;
    left: 0;
    margin-left: ${INPUT_PADDING / 16}rem;
    margin-right: ${INPUT_PADDING_RIGHT / 16}rem;
    position: absolute;
    width: -webkit-fill-available;
  }

  ${media.medium} {
    ::after {
      margin-left: ${INPUT_PADDING_MEDIUM / 16}rem;
      margin-right: ${INPUT_PADDING_RIGHT_MEDIUM / 16}rem;
    }
  }
`
