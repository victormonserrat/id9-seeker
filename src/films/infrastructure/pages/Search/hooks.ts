import { useRouter } from 'next/router'
import { useCallback } from 'react'

export const useFilmSearch = () => {
  const router = useRouter()

  const handleSearchChange = useCallback(
    async (value: string) => {
      router.replace(`?search=${value}`)
    },
    [router],
  )

  return { handleSearchChange }
}
