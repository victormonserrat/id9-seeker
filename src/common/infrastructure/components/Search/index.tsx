import { memo } from 'react'

import { useSearch } from './hooks'
import { Button, Component, Input, InputContainer } from './styles'
import { SearchProps } from './types'

const Search = ({
  className,
  disabled,
  onChange,
  value,
  ...props
}: SearchProps) => {
  const { handleChange, handleSubmit, inputValue } = useSearch({
    onChange,
    value,
  })

  return (
    <Component
      className={className}
      noValidate
      onSubmit={handleSubmit}
      role="search"
    >
      <InputContainer>
        <Input
          autoCapitalize="none"
          autoComplete="off"
          autoCorrect="off"
          disabled={disabled}
          name="search"
          onChange={handleChange}
          placeholder="Search"
          role="searchbox"
          spellCheck={false}
          type="text"
          value={inputValue}
          {...props}
        />
      </InputContainer>
      <Button disabled={disabled} type="submit">
        Search
      </Button>
    </Component>
  )
}

export default memo(Search)

export type { SearchProps }
