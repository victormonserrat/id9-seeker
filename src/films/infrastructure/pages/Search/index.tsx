import Head from 'next/head'
import Link from 'next/link'
import { useRouter } from 'next/router'

import Search from '~/common/infrastructure/components/Search'
import FilmCard from '~/films/infrastructure/components/Card'

import { useFilmSearch } from './hooks'
import { Anchor, Container } from './styles'
import { FilmSearchProps } from './types'

const FilmSearch = ({ films }: FilmSearchProps) => {
  const { handleSearchChange } = useFilmSearch()
  const router = useRouter()

  return (
    <>
      <Head>
        <title>ID9 seeker</title>
        <meta name="description" content="Star Wars films search engine" />
      </Head>
      <Container>
        <Search
          value={router.query.search?.toString()}
          onChange={handleSearchChange}
          placeholder="Search films by title, characters or planets"
        />
        {films.map((props) => (
          <Link href={props.href} key={props.id}>
            <Anchor>
              <FilmCard {...props} />
            </Anchor>
          </Link>
        ))}
      </Container>
    </>
  )
}

export default FilmSearch
