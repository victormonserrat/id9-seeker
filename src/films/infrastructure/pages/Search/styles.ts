import styled from 'styled-components'

import media from '~/common/infrastructure/styles/media'
import { Title } from '~/films/infrastructure/components/Card/styles'

export const Container = styled.div`
  display: grid;
  gap: 1rem;
  margin: auto;
  padding: 2rem 1rem;

  ${media.medium} {
    gap: 1.5rem;
    max-width: ${({ theme }) => theme.breakpoints.extralarge / 16}rem;
    padding: 3rem 1.5rem;
  }
`

export const Anchor = styled.a`
  :hover {
    ${Title} {
      color: ${({ theme }) => theme.colors.films.card.title.hover};
    }
  }
`
