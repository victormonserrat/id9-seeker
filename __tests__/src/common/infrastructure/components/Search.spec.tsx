import Search from '~/common/infrastructure/components/Search'
import { render } from '~/tests/utils'

describe('Search', () => {
  it('renders correctly', () => {
    const { container } = render(<Search />)

    expect(container).toMatchSnapshot('Regular')
  })

  it('renders disabled correctly', () => {
    const { container } = render(<Search disabled />)

    expect(container).toMatchSnapshot('Disabled')
  })
})
