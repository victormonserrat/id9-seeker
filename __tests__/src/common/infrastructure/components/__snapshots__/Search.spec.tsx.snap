// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly: Regular 1`] = `
<div>
  <form
    class="sc-acb5f70f-1 dBpRac"
    novalidate=""
    role="search"
  >
    <div
      class="sc-acb5f70f-3 kZyzbJ"
    >
      <input
        autocapitalize="none"
        autocomplete="off"
        autocorrect="off"
        class="sc-acb5f70f-2 fGCZhr"
        name="search"
        placeholder="Search"
        role="searchbox"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
    <button
      class="sc-acb5f70f-0 bNzZfY"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;

exports[`Search renders disabled correctly: Disabled 1`] = `
<div>
  <form
    class="sc-acb5f70f-1 dBpRac"
    novalidate=""
    role="search"
  >
    <div
      class="sc-acb5f70f-3 kZyzbJ"
    >
      <input
        autocapitalize="none"
        autocomplete="off"
        autocorrect="off"
        class="sc-acb5f70f-2 fGCZhr"
        disabled=""
        name="search"
        placeholder="Search"
        role="searchbox"
        spellcheck="false"
        type="text"
        value=""
      />
    </div>
    <button
      class="sc-acb5f70f-0 bNzZfY"
      disabled=""
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
